### PR TITLE
libnet: Make sure network names are unique

### LINKED ISSUE #18864: The docker daemon API lets you create two networks with the same name
The docker daemon API lets you create two networks with the same name. Having two networks with the same name breaks all kinds of things. It seems like the CLI client checks before creating a second network, but this is still a race condition. This should be prevented in the daemon.

```
docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.5.1
 Git commit:   a34a1d5-dirty
 Built:        Sun Nov 22 00:15:15 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.5.1
 Git commit:   a34a1d5-dirty
 Built:        Sun Nov 22 00:15:15 UTC 2015
 OS/Arch:      linux/amd64
```

### PR DESCRIPTION
**Pull request by akerouanton**
**- What I did**

Fixes #18864, #20648, #33561, #40901 (and probably more).

[This GH comment][1] makes clear network name uniqueness has never been enforced due to the eventually consistent nature of Classic Swarm datastores:

> there is no guaranteed way to check for duplicates across a cluster of docker hosts.

And this is further confirmed by other comments made by @mrjana in that same issue, eg. [this one][2]:

> we want to adopt a schema which can pave the way in the future for a completely decentralized cluster of docker hosts (if scalability is needed).

This decentralized model is what Classic Swarm was trying to be. It's been superseded since then by Docker Swarm, which has a centralized control plane.

To circumvent this drawback, the `NetworkCreate` endpoint accepts a `CheckDuplicate` flag. However it's not perfectly reliable as it won't catch concurrent requests.

Due to this design decision, API clients like Compose have to implement workarounds to make sure names are really unique (eg. docker/compose#9585). And the daemon itself has seen a string of issues due to that decision, including some that aren't fixed to this day (for instance moby/moby#40901):

> The problem is, that if you specify a network for a container using the ID, it will add that network to the container but it will then change it to reference the network by using the name.

To summarize, this "feature" is broken, has no practical use and is a source of pain for Docker users and API consumers. So let's just remove it for _all_ API versions.

[1]: https://github.com/moby/moby/issues/18864#issuecomment-167201414
[2]: https://github.com/moby/moby/issues/18864#issuecomment-167202589

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: integration/network/network_test.go [hunk 1]
```diff
@@ -249,3 +250,33 @@ func TestDefaultNetworkOpts(t *testing.T) {
...
+
+	c := d.NewClientT(t)
+	defer c.Close()
+	ctx := context.Background()
+
+	var wg sync.WaitGroup
+	var errors []error
+	var mu sync.Mutex
+
+	for i := 0; i < 2; i++ {
```
Comment (akerouanton) [inline integration/network/network_test.go:267]: Huh looks like I forgot a stash :see_no_evil:

File: docs/api/version-history.md [hunk 2]
```diff
@@ -38,6 +38,9 @@ keywords: "API, Docker, rcli, REST, documentation"
   specifications directories. The use of the applied setting requires the daemon
   to have expermental enabled, and for non-experimental daemons an empty list is
   always returned.
+* `POST /networks/create` with a duplicated name now fails systematically. As
+  such, the `CheckDuplicate` field is now deprecated. Note that this change is
+  _unversioned_ and applied to all API versions.
```
Comment (akerouanton) [inline docs/api/version-history.md:43]:
```suggestion
  _unversioned_ and applied to all API versions on daemon that support version 1.44.
```

Comment (akerouanton) [reply docs/api/version-history.md:43]: Should also leave yaml files untouched except for v1.44.

**Comment by akerouanton** [comment]
CI failed on a unrelated flake:

```
=== Failed
=== FAIL: amd64.integration.daemon TestLiveRestore/volume_references/local_volume_with_mount_options (2.54s)
    daemon_test.go:513: assertion failed: error is not nil: Error response from daemon: remove test-live-restore-volume-references-local: volume has active mounts
        --- FAIL: TestLiveRestore/volume_references/local_volume_with_mount_options (2.54s)

=== FAIL: amd64.integration.daemon TestLiveRestore/volume_references (18.62s)
    --- FAIL: TestLiveRestore/volume_references (18.62s)

=== FAIL: amd64.integration.daemon TestLiveRestore (0.00s)
```

File: api/types/types.go [hunk 3]
```diff
@@ -443,13 +443,7 @@ type EndpointResource struct {
...
 type NetworkCreate struct {
-	// Check for networks with duplicate names.
-	// Network is primarily keyed based on a random ID and not on the name.
-	// Network name is strictly a user-friendly alias to the network
-	// which is uniquely identified using ID.
-	// And there is no guaranteed way to check for duplicates.
-	// Option CheckDuplicate is there to provide a best effort checking of any networks
-	// which has the same name but it is not guaranteed to catch all name collisions.
+	// Deprecated: CheckDuplicate is now always enabled.
 	CheckDuplicate bool
```
Comment (thaJeztah) [inline api/types/types.go:447]: Wondering if we should make the field `omitempty` 🤔

File: client/network_create_test.go [hunk 4]
```diff
@@ -53,10 +53,9 @@ func TestNetworkCreate(t *testing.T) {
 	}
 
 	networkResponse, err := client.NetworkCreate(context.Background(), "mynetwork", types.NetworkCreate{
```
Comment (thaJeztah) [inline client/network_create_test.go:55]: We need to update `client.NetworkCreate` to always send `CheckDuplicate: true` when creating an API request to an older API version.

Otherwise, older API's without this patch will use `CheckDuplicate: false` (on the daemon side), which would allow for duplicate networks to be created without error.

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: client/network_create.go [hunk 5]
```diff
@@ -13,6 +13,9 @@ func (cli *Client) NetworkCreate(ctx context.Context, name string, options types
 		NetworkCreate: options,
 		Name:          name,
 	}
+	// CheckDuplicate is always set to true to make sure clients talking to old daemons will still get an error
+	// about duplicate networks.
+	networkCreateRequest.CheckDuplicate = true //nolint:staticcheck // ignore SA1019
```
Comment (thaJeztah) [inline client/network_create.go:18]: Ah, sorry, perhaps I should've mentioned more clearly; should we make this based on API version?

- If API-version < XX, then set it to `true`
- For higher API versions, the daemon already has it as default, so we no longer need to send it

Having it gated by version also helps us find back these kind of things in future (i.e., when we deprecate API < vXX)


I would also suggest including the API version in the `// ignore ..` part of the comment (also to help find back "since when").

File: api/types/types.go [hunk 6]
```diff
@@ -443,14 +443,8 @@ type EndpointResource struct {
...
-	// Check for networks with duplicate names.
-	// Network is primarily keyed based on a random ID and not on the name.
-	// Network name is strictly a user-friendly alias to the network
-	// which is uniquely identified using ID.
-	// And there is no guaranteed way to check for duplicates.
-	// Option CheckDuplicate is there to provide a best effort checking of any networks
-	// which has the same name but it is not guaranteed to catch all name collisions.
-	CheckDuplicate bool
+	// Deprecated: CheckDuplicate is now always enabled.
+	CheckDuplicate bool `json:",omitempty"`
```
Comment (thaJeztah) [inline api/types/types.go:447]: See my other comment. I should've mentioned that in my original comments (sorry!) perhaps mention the API version in the deprecation message, and explain that it's still set by the _Client_ when connecting to older API versions;
https://github.com/moby/moby/blob/f8a4b6755d62fb3fdcfe6d3adff5756123080c37/api/types/system/info.go#L89

File: client/network_create.go [hunk 7]
```diff
@@ -13,6 +14,10 @@ func (cli *Client) NetworkCreate(ctx context.Context, name string, options types
 		NetworkCreate: options,
 		Name:          name,
 	}
+	if versions.LessThan(cli.version, "1.44") {
```
Comment (thaJeztah) [inline client/network_create.go:17]: Note to self; I'm actually wondering now if these checks we have in the client _work_, as we moved negotiating the API version to be as late as possible, which (I _think_) in this case _could_ mean that we postpone it to `cli.post` (which calls `cli.sendRequest`, which calls `cli.getAPIPath` (which handles the negotiation)) 🙈

Comment (thaJeztah) [reply client/network_create.go:17]: (It's probably fine to keep this as-is for now, as it looks like it's something to fix for other parts as well)

File: hack/make/test-docker-py [hunk 8]
```diff
@@ -24,7 +24,8 @@ source hack/make/.integration-test-helpers
...
 # TODO re-enable test_create_network_ipv6_enabled after we updated to a version of docker-py with https://github.com/docker/docker-py/pull/3169
 # TODO re-enable test_create_with_ipv6_address after we updated to a version of docker-py with https://github.com/docker/docker-py/pull/3169
-: "${PY_TEST_OPTIONS:=--junitxml=${DEST}/junit-report.xml --deselect=tests/integration/api_container_test.py::AttachContainerTest::test_attach_no_stream --deselect=tests/integration/api_container_test.py::AttachContainerTest::test_run_container_reading_socket_ws --deselect=tests/integration/api_container_test.py::CreateContainerTest::test_create_with_restart_policy --deselect=tests/integration/errors_test.py::ErrorsTest::test_api_error_parses_json --deselect=tests/integration/api_network_test.py::TestNetworks::test_connect_with_ipv6_address --deselect=tests/integration/api_network_test.py::TestNetworks::test_create_network_ipv6_enabled --deselect=tests/integration/api_network_test.py::TestNetworks::test_create_with_ipv6_address}"
+# TODO re-enable test_create_check_duplicate after we updated to a version of docker-py with https://github.com/docker/docker-py/pull/3170
```
Comment (thaJeztah) [inline hack/make/test-docker-py:27]: oh! this reminds me that there were two tests still failing on "latest" `docker-py`. I think we missed two tests that still needed updating; https://github.com/moby/moby/pull/46368#issuecomment-1700769068

File: integration/network/network_test.go [hunk 9]
```diff
@@ -258,3 +258,21 @@ func TestDefaultNetworkOpts(t *testing.T) {
 		})
 	}
 }
+
+func TestForbidDuplicateNetworkNames(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows")
```
Comment (thaJeztah) [inline integration/network/network_test.go:263]: Do you know why this didn't fail on Windows?

Comment (akerouanton) [reply integration/network/network_test.go:263]: I force pushed my commit to check what's the error, and here it is:

```
=== Failed
=== FAIL: github.com/docker/docker/integration/network TestForbidDuplicateNetworkNames (0.01s)
    network_test.go:266: [d46d27c88db30] failed to start daemon with arguments [--data-root D:\a\moby\moby\go\src\github.com\docker\docker\bundles\tmp\TestForbidDuplicateNetworkNames\d46d27c88db30\root --exec-root C:\Users\RUNNER~1\AppData\Local\Temp\dxr\d46d27c88db30 --pidfile D:\a\moby\moby\go\src\github.com\docker\docker\bundles\tmp\TestForbidDuplicateNetworkNames\d46d27c88db30\docker.pid --userland-proxy=true --containerd-namespace d46d27c88db30 --containerd-plugins-namespace d46d27c88db30p --containerd /var/run/docker/containerd/containerd.sock --host unix://C:\Users\RUNNER~1\AppData\Local\Temp\docker-integration\d46d27c88db30.sock --debug] : protocol not available
```

So, it's totally unrelated to this change. I'll create a follow-up PR addressing that.

Comment (thaJeztah) [reply integration/network/network_test.go:263]: Interesting; thanks!

(mostly commented because we have many tests with a `skip`, and for some nobody even remembers "why" 😂)

**Comment by thaJeztah** [comment]
🙈 I think one of your other PRs conflicted; looks like this needs a rebase now (changes LGTM though)

**Review by thaJeztah — APPROVED** [review]
LGTM

File: client/network_create.go (hunk 7 above)
Comment (thaJeztah) [reply client/network_create.go:17]: opened a PR to fix this issue;

- https://github.com/moby/moby/pull/46463

**Review by neersighted — APPROVED** [review]
